### PR TITLE
[Core][CoreWorker ThreadSafety 1/n] Ensure global_worker_ is protected by mutex

### PR DESCRIPTION
**Pull request by scv119**
Theses are a serial of PRs to fix core_worker thread safety issues; mostly these issues are exposed during core_worker shutdown. The high level plan works like following:

1. ensure CoreWorkerProcess, CoreWorkerProcess singleton, and CoreWorker are thread safe.
2. in the client (python code), tolerate the case where CoreWorker has been already shutdown when CCoreWorkerProcess.GetCoreWorker() is called.
 
This PR ensure CoreWorkerProcess is thread safe, specifically, 
In CoreWorkerProcess, access to global_worker_ is not thread safe. Ensure it's guarded by the mutex_ thus thread safe.

### CODE REVIEW

**Comment by rkooo567** [comment]
oh wow. it is surprising this is not broken very frequently

File: src/ray/core_worker/core_worker.h [hunk 1]
```diff
@@ -299,18 +299,20 @@ class CoreWorkerProcess {
...
+  std::shared_ptr<CoreWorker> CreateWorker() LOCKS_EXCLUDED(mutex_);
 
   /// Remove an existing `CoreWorker` instance.
   ///
   /// \param[in] The existing `CoreWorker` instance.
   /// \return Void.
-  void RemoveWorker(std::shared_ptr<CoreWorker> worker) LOCKS_EXCLUDED(worker_map_mutex_);
+  void RemoveWorker(std::shared_ptr<CoreWorker> worker) LOCKS_EXCLUDED(mutex_);
+
+  std::shared_ptr<CoreWorker> GetGlobalWorker() LOCKS_EXCLUDED(mutex_);
```
Comment (ericl) [inline src/ray/core_worker/core_worker.h:315]: Comment?

**Review by ericl — APPROVED** [review]

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -103,10 +103,11 @@ void CoreWorkerProcess::Shutdown() {
   }
   RAY_CHECK(core_worker_process->options_.worker_type == WorkerType::DRIVER)
       << "The `Shutdown` interface is for driver only.";
-  RAY_CHECK(core_worker_process->global_worker_);
-  core_worker_process->global_worker_->Disconnect();
-  core_worker_process->global_worker_->Shutdown();
-  core_worker_process->RemoveWorker(core_worker_process->global_worker_);
+  auto global_worker = core_worker_process->GetGlobalWorker();
+  RAY_CHECK(global_worker);
+  global_worker->Disconnect();
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:108]: These global worker methods are all thread-safe right?

**Review by rkooo567 — APPROVED** [review]
Nice pattern! Btw, we need something similar for python code too.. we are accessing the global worker directly in python (which is bad for multi threaded scenario)

File: src/ray/core_worker/core_worker.cc (hunk 2 above)
Comment (scv119) [reply src/ray/core_worker/core_worker.cc:108]: Not yet. they will we

**Pull request merged**